keps/sig-auth/3130-kms-observability/README.md

```diff
@@ -0,0 +1,241 @@
+
+```go
+type EncryptRequest struct {
+	// UID is a unique identifier for the request.
+	UID *string `protobuf:"bytes,3,opt,name=uid,proto3" json:"uid,omitempty"`
+	// Version of the KMS plugin API.
+	Version string `protobuf:"bytes,1,opt,name=version,proto3" json:"version,omitempty"`
+	// The data to be encrypted.
+	Plain                []byte   `protobuf:"bytes,2,opt,name=plain,proto3" json:"plain,omitempty"`
+	XXX_NoUnkeyedLiteral struct{} `json:"-"`
+	XXX_unrecognized     []byte   `json:"-"`
+	XXX_sizecache        int32    `json:"-"`
+}
+```
+
+```go
+type DecryptRequest struct {
+	// UID is a unique identifier for the request.
+	UID *string `protobuf:"bytes,3,opt,name=uid,proto3" json:"uid,omitempty"`
+	// Version of the KMS plugin API.
+	Version string `protobuf:"bytes,1,opt,name=version,proto3" json:"version,omitempty"`
+	// The data to be decrypted.
+	Cipher               []byte   `protobuf:"bytes,2,opt,name=cipher,proto3" json:"cipher,omitempty"`
+	XXX_NoUnkeyedLiteral struct{} `json:"-"`
+	XXX_unrecognized     []byte   `json:"-"`
+	XXX_sizecache        int32    `json:"-"`
+}
+```
+
+The UID generated in the kube-apiserver will be used:
+
+1. For logging in the kube-apiserver. All envelope operations to the kms-plugin will be logged with the corresponding UID.
+   1. The UID will be logged using a wrapper in the kube-apiserver to ensure that the UID is logged in the same format and is always logged.
+   2. In addition to the UID, the kube-apiserver will also log non-sensitive metadata such as name, namespace and GroupVersionResource of the object that triggered the envelope operation.
+2. Sent to the kms-plugin as part of the `EncryptRequest` and `DecryptRequest` structs.
+
+### Test Plan
+
+Unit tests covering:
+
+1. Generation of UID for each envelope operation
+
+Integration test covering:
+
+1. Logging of UID in kube-apiserver
+2. UID in the `EncryptRequest` and `DecryptRequest`
+3. UID set to nil in the `EncryptRequest` and `DecryptRequest` when the feature gate is disabled
+   1. Confirm this results in byte equivalent data on the wire when compared to a 1.23 API server.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+-->
+
+- Feature gate
+  - Feature gate name: `KMSUID` (enabled by default)
+  - Components depending on the feature gate:
+    - kube-apiserver
+
+```go
+FeatureSpec{
+	Default: true,  // enabled by default
+	LockToDefault: false, // can be disabled if needed
+	PreRelease: featuregate.GA, // gate will dropped in 1.25
+}
+```
+
+The introduction of the new UID field is backward compatible and considered a minor change. For this reason the feature gate is enabled by default and we are starting with Stable. The pointer field plus feature gate is added as a temporary safeguard in case there are issues with the inclusion of a new field.
```

Comment:
Being backwards compatibility isn't a sufficient reason to start as stable.  All APIs are required to be backwards compatible.  I think this should follow a normal feature lifecycle.

Reply:
I'm a little skeptical as well - it's a new field, and it's adding a new concept to a stable API.  That's enough to default to normal lifecycle.  The change to the behavior surface area of an API is usually the *reason* we go through normal alpha->stable lifecycle, because once it's done we can't take it back (because stable apis are stable).

Reply:
> I'm a little skeptical as well - it's a new field, and it's adding a new concept to a stable API.

aside: the KMS gRPC API is one of our perma beta APIs 😭